src/Workspaces/Core/Portable/Progress/CodeAnalysisProgress.cs

```diff
@@ -0,0 +1,147 @@
+    /// available.  This should be used sparingly.  Locations that currently do not show progress should ideally be
+    /// migrated to ones that do so that long running operations are visible to the user in a coherent fashion.
+    /// </summary>
+    internal static readonly IProgress<CodeAnalysisProgress> None = NullProgress<CodeAnalysisProgress>.Instance;
+
+    internal bool ClearValue { get; init; }
+    internal int? CompleteItemValue { get; init; }
+    internal int? IncompleteItemsValue { get; init; }
+    internal string? DescriptionValue { get; init; }
+
+    /// <summary>
+    /// When passed to an appropriate <see cref="IProgress{T}"/>, will updates the UI showing the progress of the
+    /// current operation to the specified <paramref name="description"/>.
+    /// </summary>
+    /// <example>
+    /// progress.Report(CodeAnalysisProgress.Description("Renaming files"));
+    /// </example>
+    public static CodeAnalysisProgress Description(string description)
+        => new() { DescriptionValue = description ?? throw new ArgumentNullException(nameof(description)) };
+
+    /// <summary>
+    /// When passed to an appropriate <see cref="IProgress{T}"/>, will add the requested number of incomplete items to
+    /// the UI showing the progress of the current operation.  This is commonly presented with a progress bar.  An
+    /// optional <paramref name="description"/> can also be provided to update the UI accordingly (see <see
+    /// cref="Description"/>).
+    /// </summary>
+    /// <param name="count">The number of incomplete items left to perform.</param>
+    /// <param name="description">Optional description to update the UI to.</param>
+    /// <example>
+    /// progress.Report(CodeAnalysisProgress.AddIncompleteItems(20));
+    /// </example>
+    public static CodeAnalysisProgress AddIncompleteItems(int count, string? description = null)
+        => new()
+        {
+            IncompleteItemsValue = count >= 0 ? count : throw new ArgumentOutOfRangeException(nameof(count)),
+            DescriptionValue = description,
+        };
+
+    /// <summary>
+    /// When passed to an appropriate <see cref="IProgress{T}"/>, will indicate that some items of work have
+    /// transitioned from being incomplete (see <see cref="AddIncompleteItems"/> to complete.  This is commonly
+    /// presented with a progress bar. An optional <paramref name="description"/> can also be provided to update the UI
+    /// accordingly (see <see cref="Description"/>).
+    /// </summary>
+    /// <param name="count">The number of items that were completed. Must be greater than or equal to 1.</param>
+    /// <param name="description">Optional description to update the UI to.</param>
+    /// <example>
+    /// progress.Report(CodeAnalysisProgress.CompleteItem());
+    /// </example>
+    public static CodeAnalysisProgress CompleteItems(int count, string? description = null)
```

Comment:
honestly, i think that's a bit overwrought :)   Like, you just do two messages:

```
.Report(CodeAnalysisProgress.AddIncompleteItems(6));
.Report(CodeAnalysisProgress.CompleteItems(3));
```

Note: i would find it very weird to ever be in this state.  because you tell about the work you have *before* you do it.  So there's always a split here.

Reply:
OK, I can buy that (I'm not fully convinced that incomplete items couldn't come in after the initial incomplete items and at the same time as some completed items, but it's not a big deal). 

Also, the naming seems incongruent (AddIncompleteItems/IncompleteItems), but I think I've pushed back on that about as much as I should (I only pushed back this much as it's a public facing type)

Reply:
i don't mind 'add' for both.  it woudl make sense.   :)